src/Type/Php/MinMaxFunctionReturnTypeExtension.php

```diff
@@ -117,6 +101,53 @@ public function getTypeFromFunctionCall(FunctionReflection $functionReflection,
 		);
 	}
 
+	private function processArrayType(string $functionName, Type $argType): Type
+	{
+		$constArrayTypes = TypeUtils::getOldConstantArrays($argType);
+		if (count($constArrayTypes) > 0) {
+			$resultTypes = [];
+			foreach ($constArrayTypes as $constArrayType) {
+				$isIterable = $constArrayType->isIterableAtLeastOnce();
+				if ($isIterable->no()) {
+					$resultTypes[] = new ConstantBooleanType(false);
+					continue;
+				}
+				$iterableValueType = $constArrayType->getIterableValueType();
+				$argumentTypes = [];
+				if (!$isIterable->yes()) {
+					$argumentTypes[] = new ConstantBooleanType(false);
+				}
+
+				if ($iterableValueType instanceof UnionType) {
```

Comment:
I suspect this if/else is no longer necessary.

Reply:
Can you explain how you want the code changed @ondrejmirtes ? Because I didn't succeed having my test green as soon as I touch this if/else.

Reply:
Any code that does `instanceof *Type` is suspicious to me. Just try deleting this piece of code and see what tests fail. And try fixing them in another way.

Recently we've been adding specific methods on Type. It also means that if you rebase on top of 1.9.x, some methods you're currently using will show up as deprecated.

Reply:
Do i need to retarget to 1.9.x ?

Reply:
Would be ideal, otherwise the build will fail after the merge.

Reply:
`TypeUtils::getOldConstantArrays()` is the one that got deprecated, it would be better to use `Type::getConstantArrays()` on 1.9.x. I want to get rid of any of those, but that's probably not doable everywhere yet :)

Reply:
About the 
```
if ($iterableValueType instanceof UnionType) {
     foreach ($iterableValueType->getTypes() as $innerType) {
          $argumentTypes[] = $innerType;
     }
} else {
     $argumentTypes[] = $iterableValueType;
}
```
code

If I only keep the if part, I get an error, because `getTypes` method only exists in UnionType.
If I only keep the else part, the following test is failing
```
assertType('9', max([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]));
```
because, we're adding to the arguments the iterableValueType `0|1|2|...|8|9` instead of adding all the value types `0`, `1`, ..., `8`, `9` separately.

Reply:
Only slightly related note: I'm still thinking about easier Union type iteration and creating something like a Type::getUnionedTypes() is still in the back of my head. works mostly also as flattenTypes() successor, but it is hiiighly specific

Reply:
> Only slightly related note: I'm still thinking about easier Union type iteration and creating something like a Type::getUnionedTypes() is still in the back of my head. works mostly also as flattenTypes() successor, but it is hiiighly specific

This would be a great solution of what is asked to me (not using `instanceof UnionType` anymore).
Currently, I don't see how I avoid this instanceof check. But maybe you have an idea how I should rewrite the code atm ?

Reply:
not sure, the thing I mentioned can still be done later, I'm still not sure about it. I also don't like the name yet. because sometimes I explicitly want to iterate only union types and sometimes I want to iterate whatever this type is and "spread" the union types, basically e.g. a single IntegerType would be returned as either `[]` or `[IntegerType]` 🤔
sorry, I don't want to drift off here too much, I'll be silent again :)

Reply:
For the record @ondrejmirtes, the same check exists here: https://github.com/phpstan/phpstan-src/pull/1795/files#diff-b1a10b0e3c0ccbea063bae8da98c7f26d862672ddc4e2992db23ba79c5b07121R84-R90

Reply:
No, flattenTypes is 💩 , more likely we need a specific method for min/max, so that IntegerRangeType can also answer correctly.

Reply:
that, of course, would be the best and cleanest solution!

Reply:
> No, flattenTypes is 💩 , more likely we need a specific method for min/max, so that IntegerRangeType can also answer correctly.

So I need to add a new method in the Type interface instead of doing
```
if ($iterableValueType instanceof UnionType) {
    foreach ($iterableValueType->getTypes() as $innerType) {
        $argumentTypes[] = $innerType;
    }
} else {
    $argumentTypes[] = $iterableValueType;
}
```
?

This doesn't sound like an easy things to do. Shouldn't be done in another PR in order to not delay the bugfix since the `instanceof UnionType` code was already here ?

Reply:
"The window was already broken, I don't need to fix it now."

Sorry, this isn't how I imagine and manage quality software development. The fix needs to make sense and is easier to do once the knowledge of the specific area is already in your headspace.

Reply:
> Sorry, this isn't how I imagine and manage quality software development. The fix needs to make sense and is easier to do once the knowledge of the specific area is already in your headspace.

I'll be more than happy to find another way to write this code but I'll need more help then. Because I didn't understood everything during this discussion. You said "Recently we've been adding specific methods on Type" and "we need a specific method for min/max" so do you want me to add a new method to types ? What should it does ?

Reply:
@herndlm can help answering that but usually you check recent merged PR history and also the above discussion to get the idea.

Reply:
I saw Pr like https://github.com/phpstan/phpstan-src/pull/1869 which introduced a new method.

My main concern was about the method to introduce.
I thought about introducing a `getTypes` method to all the Type, in order to write:
```
foreach ($iterableValueType->getTypes() as $innerType) {
    $argumentTypes[] = $innerType;
}
```
But i'm not sure you will like this solution.

You said "we need a specific method for min/max" but at first sight that would mean reproducing/duplicating part of the logic from MinMaxFunctionReturnTypeExtension::processType/compareTypes in every types.

Reply:
sorry, I didn't fully dive into this yet, but it sounds like you can add a `Type::max(Types ...$types): Type` and min methods. moving it to type often could mean slight duplications, but the core logic for arrays should then be split into `ArrayType` and `ConstantArrayType` and we automatically add support for unions and e.g. don't need `Type::getConstantArrays()` & co anymore. The other types are mostly covered via the array traits and we need to think explicitly how the array accessory types like `HasOffsetType` should behave. it is possible then to simply add code for them as well via the new methods.

I didn't think through yet how you can solve the iterableValueType problem that seems to be here. because it sounds like it would just be moved somewhere else. but you have then `max` and `min` methods on type, so maybe that helps somehow. since max and min don't only handle arrays, it means most likely that other types need to implement too somehow, I have no clue yet how, so I'm not sure if that helps much here. Maybe you also need `minArray()` and `maxArray()` methods on type and want to keep the rest here. again, I didn't dive into this, sorry

Reply:
> how you can solve the iterableValueType problem

The previous logic was
- If it's an Array
  - if the IterableValueType is an UnionType, iterate over the types.
  - else take the iterableValueType like this.

The logic I neeeded to implements in the ReturnTypeExtension is
- If it's a ConstantArray
  - if the IterableValueType is an UnionType, iterate over the types.
  - else take the iterableValueType like this.
 - If it's just an Array
   -  take the iterableValueType like this.

I just discovered in case it's a ConstantArray, it could just be replaced by the `getValueTypes` method !